tests/generators/cpp/src/test.cpp

```diff
@@ -0,0 +1,353 @@
+
+      THEN("default is returned") { CHECK(value == Fixture::kDefaultValue); }
+    }
+  }
+}
+
+TEMPLATE_TEST_CASE("Property setters", "", openassetio::Bool, openassetio::Int, openassetio::Float,
+                   openassetio::Str) {
+  using openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait;
+
+  using PropertyType = TestType;  // Catch2-injected template param.
+  using Fixture = PropertyFixture<PropertyType>;
+
+  GIVEN("an AllPropertiesTrait view of a blank TraitsData") {
+    const openassetio::TraitsDataPtr traitsData = openassetio::TraitsData::make();
+    Fixture fixture{AllPropertiesTrait{traitsData}};
+
+    WHEN("property is set") {
+      fixture.setProperty(Fixture::kExpectedValue);
+
+      THEN("TraitsData contains expected value") {
+        openassetio::trait::property::Value value;
+        [[maybe_unused]] const bool hasProp =
+            traitsData->getTraitProperty(&value, AllPropertiesTrait::kId, Fixture::kPropertyKey);
+
+        const PropertyType actualValue = std::get<PropertyType>(value);
+
+        CHECK(actualValue == Fixture::kExpectedValue);
+      }
+
+      THEN("TraitsData has been imbued") { CHECK(traitsData->hasTrait(AllPropertiesTrait::kId)); }
+    }
+  }
+}
+
+SCENARIO("Moveable property setters") {
+  using openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait;
+
+  GIVEN("an AllPropertiesTrait view of a blank TraitsData") {
+    const openassetio::TraitsDataPtr traitsData = openassetio::TraitsData::make();
+    AllPropertiesTrait trait{traitsData};
+
+    WHEN("string property is set with a moveable string") {
+      std::string value = "some string";
+      trait.setStringProperty(std::move(value));
+
+      THEN("string has been moved") {
+        // TODO(DF): There is currently no way to check that the value
+        //  was moved "all the way", e.g. check `.data()` pointer is the
+        //  same, since `TraitsData::getTraitProperty` always copies.
+        CHECK(value.empty());  // NOLINT(bugprone-use-after-move)
+      }
+    }
+  }
+}
+
+SCENARIO("LocalAndExternalTraitSpecification") {
+  GIVEN("a LocalAndExternalTraitSpecification") {
+    const auto traitsData = openassetio::TraitsData::make();
+
+    const openassetio_traitgen_test_all::specifications::test::LocalAndExternalTraitSpecification
+        specification{traitsData};
+
+    WHEN("an externally defined trait view is requested") {
+      const auto trait =
+          specification.openassetioTraitgenTestTraitsOnlyANamespaceNoPropertiesTrait();
+
+      THEN("trait view has expected type") {
+        STATIC_REQUIRE(
+            std::is_same_v<
+                std::remove_const_t<decltype(trait)>,
+                openassetio_traitgen_test_traits_only::traits::aNamespace::NoPropertiesTrait>);
+      }
+
+      // TODO(DF): The python tests check the underlying data pointer
+      //  directly, but data() is a protected method here, should it be?
```

Comment:
Feels like we need to be consistent - it's protected in Python, so be odd for it to be public in C++.

Reply:
Tom highlighted this too. See discussion in https://github.com/OpenAssetIO/OpenAssetIO-TraitGen/pull/12#discussion_r1126222850

Reply:
Removed the TODO comment in eaf62895